ctru-rs/src/error.rs

```diff
@@ -1,11 +1,45 @@
 use std::error;
 use std::ffi::CStr;
 use std::fmt;
+use std::ops::{ControlFlow, FromResidual, Try};
 
 use ctru_sys::result::{R_DESCRIPTION, R_LEVEL, R_MODULE, R_SUMMARY};
 
 pub type Result<T> = ::std::result::Result<T, Error>;
 
+#[derive(Debug, Clone, PartialEq, PartialOrd)]
+#[repr(transparent)]
+pub(crate) struct LibCtruResult(pub i32);
+
+impl Try for LibCtruResult {
+    type Output = ();
+    // This type is passed to [FromResidual::from_residual] when the LibCtruResult is an error,
+    // so this type implies "this is a result than CAN'T be `Ok`" (Infallible is the same as !)
+    // The purpose of this type is to _only_ bring information about the *Error*
+    type Residual = crate::Result<core::convert::Infallible>;
+
+    fn from_output(_: Self::Output) -> Self {
+        Self(0)
+    }
+
+    fn branch(self) -> ControlFlow<Self::Residual, Self::Output> {
+        if self.0 < 0 {
+            ControlFlow::Break(Err(self.into()))
+        } else {
+            ControlFlow::Continue(())
+        }
+    }
+}
+
+impl FromResidual for LibCtruResult {
+    fn from_residual(e: <Self as Try>::Residual) -> Self {
+        match e.err().unwrap() {
+            Error::Os(result) => Self(result),
+            _ => Self(-1),
+        }
+    }
```

Comment:
Note: if we enable the [`exhaustive_patterns`](https://github.com/rust-lang/rust/issues/51085) feature this can be:
```suggestion
    fn from_residual(e: <Self as Try>::Residual) -> Self {
        match e {
            Err(Error::Os(result)) => Self(result),
            Err(_) => Self(-1),
        }
    }
```

Reply:
@Meziu What do you think about this?

Reply:
It is too small of a change to warrant the use an unstable feature. Obviously it isn’t a dangerous feature, but I’d like to keep things tidy.